src/DomainHost.sol

```diff
@@ -0,0 +1,427 @@
+            uint256 diff = ink - art;
+            vat.suck(_vow, _vow, diff * RAY); // This needs to be done to make sure we can deduct sin[vow] and vice in the next call
+            vat.grab(ilk, address(this), address(this), _vow, 0, _int256(diff));
+        }
+
+        require(dai.transferFrom(escrow, address(this), wad), "DomainHost/transfer-failed");
+        daiJoin.join(address(this), wad);
+        int256 repay = -_int256(wad);
+        vat.frob(ilk, address(this), address(this), address(this), repay, repay);
+        vat.slip(ilk, address(this), repay);
+
+        grain -= wad;
+
+        emit Release(wad);
+    }
+
+    /// @notice Guest is pushing a surplus (or deficit)
+    /// @param _lid Local ordering id
+    /// @param wad The amount of DAI to push (or pull) [WAD]
+    function push(uint256 _lid, int256 wad) external guestOnly ordered(_lid) {
+        if (wad >= 0) {
+            dai.transferFrom(address(escrow), address(this), uint256(wad));
+            daiJoin.join(address(vow), uint256(wad));
+        } else {
+            sin += uint256(-wad);
+        }
+
+        emit Push(wad);
+    }
+
+    /// @notice Move bad debt from the remote domain into the local vow
+    /// @dev This is a potentially dangerous operation as a malicious domain can drain the entire surplus buffer
+    /// Because of this we require an authed party to perform this operation
+    function _rectify() internal auth returns (uint256 _rid, uint256 _wad) {
+        require(vat.live() == 1, "DomainHost/vat-not-live");
+
+        uint256 wad = sin;
+        require(wad > 0, "DomainHost/no-sin");
+        vat.suck(vow, address(this), wad * RAY);
+        daiJoin.exit(address(escrow), wad);
+        sin = 0;
+        
+        _rid = rid++;
+        _wad = wad;
+
+        emit Rectify(wad);
+    }
+
+    /// @notice Initiate shutdown for this domain
+    /// @dev This will trigger the end module on the remote domain
+    function _cage() internal returns (uint256 _rid) {
+        require(vat.live() == 0 || wards[msg.sender] == 1, "DomainHost/not-authorized");
+        require(live == 1, "DomainHost/not-live");
+
+        live = 0;
+
+        _rid = rid++;
+
+        emit Cage();
+    }
+
+    /// @notice Set this domain's cure value
+    /// @param _lid Local ordering id
+    /// @param value The value of the cure [RAD]
+    function tell(uint256 _lid, uint256 value) external guestOnly ordered(_lid) {
+        require(live == 0, "DomainHost/live");
+        require(!cureReported, "DomainHost/cure-reported");
+        require(_divup(value, RAY) <= grain, "DomainHost/cure-bad-value");
+
+        cureReported = true;
+        cure = value;
+
+        emit Tell(value);
+    }
+
+    /// @notice Allow DAI holders to exit during global settlement
+    /// @param usr The address to send the claim token to
+    /// @param wad The amount of gems to exit [WAD]
+    function _exit(address usr, uint256 wad) internal returns (address _usr, uint256 _wad) {
```

Comment:
`usr` seems to be an l2 address so it should be a `uint256`

Reply:
We can overload this function, but we use `bytes32` for abstract address.

Reply:
Right, I should have written `bytes32`.

Reply:
Do you want to overload in baseclass or in subclasses? My point is that methods in base class should be as general as necessary.